### PR TITLE
Simplify and fix setattr logic.

### PR DESCRIPTION
**Pull request by levskaya**
Simplifies setattr logic and allows users to set attributes inside `__post_init__` without spurious errors.

There was too much confusing special-casing going on inside setattr, this simplifies that logic and removes some needlessly irritating constraints on our users.  Added reference to `__post_init__` usage in our error docs.

### CODE REVIEW

**Review by marcvanzee — APPROVED** [review]
Very nice and useful change! One thing that would be useful to clarify somehow is in what order `setup()` and `__post_init__()` are called. I suppose `setup()` runs before `__post_init__()`, but it is not directly clear if you just look at the names of the methods.

I can imagine it is useful to add a test that sets an attribute on `self` in `__post_init__()`, which uses some information defined in `setup()`.

File: flax/errors.py [hunk 1]
```diff
@@ -423,6 +423,26 @@ def __call__(self, x):
 
     variables = SomeModule(features=3).init(random.PRNGKey(0), jnp.ones((1, )))
 
+  Alternatively, if you need to introduce canonicalization or automatically derive
+  other attributes, you can add logic to `__post_init__`:
+
+    class Foo(nn.Module):
```
Comment (jheek) [inline flax/errors.py:429]: This only works if `canonicalize(canonicalize(orig_attributes)) == canonicalize(orig_attributes)` otherwise cloning can cause side-effects. @avital has a PR that stores the initial attr values to solve this case (and enables atttribute updates in setup)

**Comment by levskaya** [comment]
actually `__post_init__` is always called -before- `setup` - it is called automatically by the `init` method generated by the dataclass decorator.  Recall that `setup` is called -lazily- only before any user-defined method or `getattr`.

File: flax/errors.py (hunk 1 above)
Comment (levskaya) [reply flax/errors.py:429]: ah right of course,  I'll axe the text in the docs for this PR then - I still think the logic of this PR is an improvement on the setattr jumble that we had before and should be compatible with any such further fixes.

Comment (levskaya) [reply flax/errors.py:429]: note that the motivating case was one of deriving other attributes from dataclass attribs rather than normalizing dataclass attribs directly.

Comment (levskaya) [reply flax/errors.py:429]: @avital - you have a link to that PR?

**Comment by lucasb-eyer** [comment]
Thanks for this quick change Anselm!
